### ISSUE TITLE
Resize maps

### ISSUE THREAD

**Issue by L89David** [ChaosInitiative/Portal-2-Community-Edition#328]
I know, this seems like something really hard to pull off, or impossible but it would be amazing.

We know that the size of a map has a limit on all axis. But what if we could resize the map without exceeding the maximum default size? This means that we could fit for example a map like "Long Fall" from sp without moving the tunnel in the entire map. Or have extremely large maps, but with not too much height.

**Comment by SpyceTewan** [comment]
Neither the compilers nor the engine like large spaces... Idk if this is really necessary...

**Comment by 20kdc** [comment]
(not a dev, but know the limitations of floats and what happens when BSP interacts with bad precision)
I suspect there might be a misunderstanding here. The map boundaries are absolute because of a mixture of floating point limitations (which affect every single datastructure that holds a position or even an indirect representation of one across the entire engine) and possible BSP compiler weirdness.
The wording "Or have extremely large maps, but with not too much height." suggests the reduction of one axis to gain room in another, but that's not actually how it works.
Simply put, it would likely be very easy to increase the limitations on map size, uniformly, but making such a thing *stable* would require switching to doubles across the entire engine, which is somewhat unlikely, might introduce issues, might make the engine slower, might break down at the GPU level, etc.

**Comment by JJL772** [comment]
>  The map boundaries are absolute because of a mixture of floating point limitations (which affect every single datastructure that holds a position or even an indirect representation of one across the entire engine) and possible BSP compiler weirdness.

The bounds are absolute because Source's BSP format still uses shorts (16-bit ints) to store coords for planes, brushes and other data. I assume this is because Valve hasn't really had a need for larger maps, and there are countless examples of this (see every map made by valve, they're all quite a bit smaller than the grid). As for floating-point limitations, maps that are 32k x 32k x 32k units are far too small to be affected by floating point precision issues. You'd need to increase the size a LOT to observe any floating point weirdness.

> Simply put, it would likely be very easy to increase the limitations on map size, uniformly, but making such a thing stable would require switching to doubles across the entire engine, which is somewhat unlikely, might introduce issues, might make the engine slower, might break down at the GPU level, etc.

With larger map sizes, you'll probably see higher memory usage on both the client and the server, and probably a slowdown in the map loading speed. Other than that though, I highly doubt that you'd see a drop in performance with larger maps, granted that they play towards the strengths of BSP (i.e. no large open spaces which BSP tends to struggle with). Backwards compatibility is a pretty big issue here though, so we'd probably need to change the file extension for the expanded/extended maps.

Regardless, I agree with Tewan. I dont think there's much of a need for this in p2ce.

**Comment by Luke18033** [comment]
So shouldn't this be closed?
